demos/jobs_for_aws/jobs_demo.c

```diff
@@ -0,0 +1,789 @@
+ */
+static void prvEventCallback( MQTTContext_t * pxMqttContext,
+                              MQTTPacketInfo_t * pxPacketInfo,
+                              MQTTDeserializedInfo_t * pxDeserializedInfo )
+{
+    uint16_t usPacketIdentifier;
+
+    ( void ) pxMqttContext;
+
+    configASSERT( pxDeserializedInfo != NULL );
+    configASSERT( pxMqttContext != NULL );
+    configASSERT( pxPacketInfo != NULL );
+
+    usPacketIdentifier = pxDeserializedInfo->packetIdentifier;
+
+    /* Handle incoming publish. The lower 4 bits of the publish packet
+     * type is used for the dup, QoS, and retain flags. Hence masking
+     * out the lower bits to check if the packet is publish. */
+    if( ( pxPacketInfo->type & 0xF0U ) == MQTT_PACKET_TYPE_PUBLISH )
+    {
+        configASSERT( pxDeserializedInfo->pPublishInfo != NULL );
+        JobsTopic_t topicType = JobsMaxTopic;
+        char * pcJobId = NULL;
+        uint16_t usJobIdLength = 0;
+        JobsStatus_t xStatus = JobsError;
+
+        LogDebug( ( "Received an incoming publish message: TopicName=%.*s",
+                    pxDeserializedInfo->pPublishInfo->topicNameLength,
+                    ( const char * ) pxDeserializedInfo->pPublishInfo->pTopicName ) );
+
+        /* Let the Device Shadow library tell us whether this is a device shadow message. */
+        xStatus = Jobs_MatchTopic( ( char * ) pxDeserializedInfo->pPublishInfo->pTopicName,
+                                   pxDeserializedInfo->pPublishInfo->topicNameLength,
+                                   democonfigTHING_NAME,
+                                   THING_NAME_LENGTH,
+                                   &topicType,
+                                   &pcJobId,
+                                   &usJobIdLength );
+
+        if( xStatus == JobsSuccess )
+        {
+            /* Upon successful return, the messageType has been filled in. */
+            if( ( topicType == JobsStartNextSuccess ) || ( topicType == JobsNextJobChanged ) )
+            {
+                /* Handler function to process payload. */
+                prvStartNextJobHandler( pxDeserializedInfo->pPublishInfo );
+            }
+            else if( topicType == JobsUpdateSuccess )
+            {
+                LogInfo( ( "Job update status request has been accepted by AWS Iot Jobs service." ) );
+            }
+            else if( topicType == JobsStartNextFailed )
+            {
+                LogWarn( ( "Request for next job description rejected: RejectedResponse=%.*s.",
+                           pxDeserializedInfo->pPublishInfo->payloadLength,
+                           ( const char * ) pxDeserializedInfo->pPublishInfo->pPayload ) );
+            }
+            else if( topicType == JobsUpdateFailed )
+            {
+                /* Set global flag to terminate demo as request for updating executing the job status
+                 * has been rejected by AWS IoT Jobs service. */
```

Comment:
```suggestion
                /* Set the global flag to terminate the demo, because the request for updating an executing the job status
                 * has been rejected by AWS IoT Jobs service. */
```